API: Allow `Stream` to be set in `PodLogOptions`

knight42 opened the pull request:
#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:

This PR adds the feature gate and API for https://github.com/kubernetes/enhancements/issues/3288 as well as validation rules to prevent inconsistent values. Changes about API, kube-apiserver and kubelet is included in this PR.

#### Special notes for your reviewer:

I would like to file another PR to update kubectl in case the volume of this PR is too large.


#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Add a `Stream` field to `PodLogOptions`, which allows clients to request certain log stream(stdout or stderr) of the container.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3288-separate-stdout-from-stderr
```

knight42 commented:
/priority important-soon
/triage accepted
/cc @SergeyKanzhelev @mrunalp
/assign @liggitt

knight42 commented:
/test pull-kubernetes-integration

knight42 commented:
gently ping @liggitt 👋 , would you mind taking a look when available?

liggitt commented:
/assign @smarterclayton 
/unassign

routing to clayton based on https://github.com/kubernetes/kubernetes/blob/master/OWNERS_ALIASES#L358-L360 and review load in https://github.com/orgs/kubernetes/projects/13?card_filter_query=assignee%3Asmarterclayton+is%3Aopen

liggitt commented:
> Subsequent PRs will add implementation.

While it can be ok to review an API PR in isolation, we avoid merging APIs ahead of implementation to prevent shipping APIs prematurely

liggitt commented:
> @liggitt Got it! Is it okay to push the implementation code to this PR or is it better to file another one?

I've seen both approaches... a separate PR makes it easier to have parallel reviews of the API and implementation by different people, a single PR can be easier to manage.

Either way, keeping the API bits and the implementation bits in separate commits is helpful.

smarterclayton commented on pkg/apis/core/types.go:
```diff
@@ -4643,6 +4643,18 @@ type Preconditions struct {
+
+const (
+	// LogStreamTypeStdout is the stream type for stdout.
+	LogStreamTypeStdout LogStreamType = "stdout"
```
I don't see mention of this in the KEP or discussion, but this is a constant in our public API and should follow constant rules.  For instance, `all` should be `All`.  While `stdout` and `stderr` are terms of art, they are generally not directly referenced via name except in documentation.  The POSIX spec *does* reference them in their lower case form (https://pubs.opengroup.org/onlinepubs/9699919799/functions/stdin.html) but does not define how to use them.  I would slightly lean towards `stderr` over `Stderr` or `StdErr` or `STDERR`, but will get input from other reviewers.

So go ahead and change `all` to `All` unless there is previous discussion justifying the non-standard constant, and I may come back and suggest a change to the other constant values depending on other opinions.

smarterclayton commented on pkg/apis/core/types.go:
```diff
@@ -4814,7 +4835,7 @@ type ObjectReference struct {
 // LocalObjectReference contains enough information to let you locate the referenced object inside the same namespace.
 type LocalObjectReference struct {
-	//TODO: Add other useful fields.  apiVersion, kind, uid?
+	// TODO: Add other useful fields.  apiVersion, kind, uid?
```
Was this change gofmt induced?  Or accidental?

knight42 replied:
Yeah this kind of changes is caused by the formatter in goland.

smarterclayton commented on pkg/apis/core/v1/conversion.go:
```diff
@@ -540,3 +541,12 @@ func Convert_core_PersistentVolumeSpec_To_v1_PersistentVolumeSpec(in *core.Persi
 }
+
+func Convert_Slice_string_To_Pointer_v1_LogStreamType(in *[]string, out **v1.LogStreamType, s conversion.Scope) error {
+	if len(*in) == 0 {
```
Was this not auto generated?

knight42 replied:
I am not sure, if I don't provide this conversion function, a `FIXME` comment would be generated inside the `autoConvert_url_Values_To_v1_PodLogOptions` function, I guess the generator didn't know how to perform conversion between `[]string` and `*v1.LogStreamType`

smarterclayton replied:
Ok.

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -5488,6 +5500,15 @@ type PodLogOptions struct {
 	// +optional
 	InsecureSkipTLSVerifyBackend bool `json:"insecureSkipTLSVerifyBackend,omitempty" protobuf:"varint,9,opt,name=insecureSkipTLSVerifyBackend"`
+
+	// Return the specific log stream. Defaults to "all".
```
Let's improve this godoc by describing the action in greater detail:

```
Specify which container log streams to return to the client. By default both stdout and stderr are returned interleaved.  <describe what happens on Windows>.
Acceptable values are:
This is an alpha field ....
```

knight42 replied:
> <describe what happens on Windows>.

I think this feature should make no difference on windows 🤔

smarterclayton replied:
Do windows containers have stderr and stdout?

knight42 replied:
Yeah I believe so https://docs.microsoft.com/en-us/virtualization/windowscontainers/troubleshooting#docker-container-logs IIUC.

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -5488,6 +5500,15 @@ type PodLogOptions struct {
+	// - "all": the combined stdout and stderr
+	// - "stdout": only the stdout
+	// - "stderr": only the stderr
+	// This is an alpha field and requires enabling SplitStdoutAndStderr feature gate.
```
Just to clarify here (to make sure I understand the intent in the KEP) - the API *will not* accept the empty string (which is our convention for new fields), but kubectl will accept `--stream ""` as the empty string?

debdeepsanyal replied:
Hi Clayton,
I'm new to this community and i was looking for some guidance through the process and I'll be really grateful if you take some time out to help me through and become an integral part of this community.

knight42 replied:
> but kubectl will accept --stream "" as the empty string?

Nope, the default value of flag `--stream` would be `all`, and I plan to reject the empty string.

smarterclayton replied:
Ok, so both CLI and API will reject empty string?

knight42 commented:
I decided to include changes about apiserver and kubelet in this PR, and submit another PR to update kubectl.

knight42 commented:
I have updated the description of this PR to note the changes of each primary commit to make this PR easier to review.

For the API/kube-apiserver changes, cc @smarterclayton 
For the kubelet changes, cc @mrunalp @dchen1107 @SergeyKanzhelev

smarterclayton commented on pkg/kubelet/kuberuntime/logs/logs.go:
```diff
@@ -207,10 +217,46 @@ func getParseFunc(log []byte) (parseFunc, error) {
+func extractStreamFromLog(log []byte) (runtimeapi.LogStreamType, error) {
+	var stream runtimeapi.LogStreamType
+
+	if bytes.HasPrefix(log, []byte{'{'}) {
```
Doing this check in this method is expensive (one check per line, when this should never change on a file by file perspective).  I would expect us to do this no more than once per file, at most.

What is the estimated end of life for docker log files?  Is this an option we support on the container runtime, or a consequence of dockershim support (and thus no longer needed)?

If we are still planning on supporting docker json format, then this needs to be more efficient.

knight42 replied:
> I would expect us to do this no more than once per file, at most.

Yeah I think it makes sense, I would update this function to parse once per file.

> Is this an option we support on the container runtime, or a consequence of dockershim support 

IIUC the log files is managed by the container runtime, dockershim just translate the API calls, so I guess we need to support docker json log format.

smarterclayton commented on pkg/kubelet/kuberuntime/logs/logs.go:
```diff
@@ -236,10 +283,15 @@ func newLogWriter(stdout io.Writer, stderr io.Writer, opts *LogOptions) *logWrit
 func (w *logWriter) write(msg *logMessage) error {
-	if msg.timestamp.Before(w.opts.since) {
-		// Skip the line because it's older than since
+	switch {
```
This method is already huge, and this change makes it larger.  I'd like to see this split up into two paths (if one is substantially different), with a commit moving the existing path unchanged to its own method happening separately (so we can review that independently).  This is already hard to read, complex code, and then new code is also hard to read and complex, so we need to be reducing the difficulty of review and longer term support by focusing on clarity and clear separation.

knight42 replied:
Sure! I would try to make the function easier to review

smarterclayton commented on pkg/kubelet/kuberuntime/logs/logs_test.go:
```diff
@@ -72,137 +78,307 @@ func TestLogOptions(t *testing.T) {
+				TailLines: pointer.Int64Ptr(8),
 			},
-			expected: "line1\nline2\nline3\n",
+			expectedStdout: "line1\nline2\nline3\nline5\n",
```
I did not expect the existing test to change - in fact, I expected this test to either add new cases, or check additional invariants, but not to change the behavior of `expected`.  

When the feature gate is off, this method should return exactly the same output in exactly the same order as it did before, and with the split here we can't prove that (and hence from review perspective I am not comfortable with the change).

I'm fine with adding `expectedStdout` and `expectedStderr` as *additional* fields, and leaving `expected` as-is (you can run ReadLogs twice if you have to in the inner loop).

smarterclayton commented on pkg/kubelet/kuberuntime/logs/filter_log_test.go:
```diff
@@ -0,0 +1,53 @@
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
+)
+
+func TestFilterLog(t *testing.T) {
```
I expected to see a lot more test scenarios here, including bounds and limit testing (would like all of the major branches in filter unit tested here).

smarterclayton commented on pkg/kubelet/kuberuntime/logs/filter_log.go:
```diff
@@ -0,0 +1,164 @@
+)
+
+// circularLinkedList is an append-only linked list, and behaves like a ring buffer.
+type circularLinkedList struct {
```
A new data structure needs a lot of testing.  I'd expect significant unit tests for the data structure verifying the contract.

Also, can you more clearly justify why you need a new data structure vs reusing one of the existing stdlib data structures in comments inline?  I'm hazy on why `container/ring` is not sufficient - a comment inline justifying it would be preferred.

knight42 replied:
Actually I have considered `container/ring` before, I decided to add this new data structure mainly for memory efficiency.

`container/ring` allocates all nodes on creation, this might not be sufficient. As for the current implementation, we would use the `tailLines` specified by user as the capacity of the circular list. Let's say a log file only contains 10 lines, and the `tailLines` is set to 1000 by user. `container/ring` would allocate 1000 elements on creation which is not quite memory efficient, while the added `circularLinkedList` would only create new element if necessary.

smarterclayton replied:
How many tail logs would we have to have concurrently before that is an issue?   I would suspect the per connection request overhead for those would be significantly higher than container/ring, and so if the only purpose of this data structure is to lazily allocate elements, that's hard to justify against the extra complexity here.  Do you have numbers to support why the efficiency of the ring buffer here is the key issue?

rphillips commented on pkg/kubelet/kuberuntime/logs/logs.go:
```diff
@@ -207,10 +217,46 @@ func getParseFunc(log []byte) (parseFunc, error) {
+		var obj struct {
+			Stream runtimeapi.LogStreamType `json:"stream"`
+		}
+		err := json.Unmarshal(log, &obj)
```
The json.unmarshal is potentially expensive as well. Is there a way we can do this without deserializing the entire line? It is also expensive from a GC persepective.

knight42 replied:
🤔 I would also like to avoid this deserializing, but the docker log is formatted in JSON, I don't know if there is another robust way to get the stream type from a single line of log.

dims commented:
If you still need this PR then please rebase, if not, please close the PR

knight42 commented:
> If you still need this PR then please rebase, if not, please close the PR

Sorry for the late reponse, I am getting too busy for the work of my job now. I have converted this PR to draft PR, I will get back to this when I am available to work on it.

knight42 committed:
feat(api): add field stream to PodLogOptions

Signed-off-by: Jian Zeng <anonymousknight96@gmail.com>

knight42 committed:
feat: update conversion code

Signed-off-by: Jian Zeng <anonymousknight96@gmail.com>

knight42 committed:
feat: update validation helpers

Signed-off-by: Jian Zeng <anonymousknight96@gmail.com>

knight42 committed:
feat: add feature gate

Signed-off-by: Jian Zeng <anonymousknight96@gmail.com>

knight42 committed:
feat(apiserver): set stream param in LogLocation

Signed-off-by: Jian Zeng <anonymousknight96@gmail.com>

knight42 committed:
feat(kubelet): only returns logs that match the given `stream`

Note that if `Stream` is used in conjunction with `TailLines`, the
output is the logs from the last `TailLines` that match the given
`Stream`, rather than the last `TailLines` logs of the given `Stream`.

The implementation of the latter requires parsing log entry to extract
the log stream and a buffering mechanism (the memory required is proportional to `TailLines`)
when reading container logs, which is not efficient and prone to DoS attack.

Signed-off-by: Jian Zeng <anonymousknight96@gmail.com>

knight42 commented:
/reopen

knight42 commented:
@smarterclayton @SergeyKanzhelev Hi, sorry for the late response I finally managed to rewrite the code a bit, addressing the comments, and I would like to point out that I decided to change the API semantics when `Stream` is used in conjunction with `TailLines`.

Right now, when user specifies both `Stream` and `TailLines`, the output is the logs from the last `TailLines` that match the given `Stream`, rather than the last `TailLines` logs of the given `Stream`.

For instance, let's say the logs of a container is as follows:
```
[stdout] line 1
[stderr] line 2
[stdout] line 3
[stderr] line 4
```
If a user request the container logs with `stream=stderr&tailLines=2`, the result is `line 4\n`, instead of `line 2\nline 4\n`.

The reason I made this decision is I found out the implementation of the latter requires parsing log entry to extract
the log stream and a buffering mechanism (the memory required is proportional to `TailLines`) when reading container logs, which is not efficient and prone to DoS attack.


It is worth mentioning that, dockerd takes the same approach, which can be verified by the following steps:

Run an example container:
```
$ docker run --name testlogs alpine:edge sh -c 'for i in $(seq 1 10); do if [ $(expr $i % 2) -eq 0 ]; then echo stdout $i; else echo stderr $i >&2; fi; done'
stderr 1
stdout 2
stderr 3
stdout 4
stderr 5
stdout 6
stderr 7
stdout 8
stderr 9
stdout 10
```

Since `docker` cli does not allow us to specify the stream, we need to call the dockerd API directly. I wrote a small program to return the `stdout` stream of container logs with user-specified `TailLines`:
<details>
<summary>$ cat main.go</summary>

```go
package main

import (
	"context"
	"flag"
	"os"

	"github.com/docker/docker/api/types"
	"github.com/docker/docker/client"
	"github.com/docker/docker/pkg/stdcopy"
)

func main() {
	var tail string
	flag.StringVar(&tail, "n", "1", "tail last n lines")
	flag.Parse()

	cli, err := client.NewClientWithOpts(client.WithAPIVersionNegotiation(), client.WithHostFromEnv())
	if err != nil {
		panic(err)
	}
	reader, err := cli.ContainerLogs(context.TODO(), "testlogs", types.ContainerLogsOptions{
		ShowStdout: true,
		Tail:       tail,
	})
	if err != nil {
		panic(err)
	}
	_, err = stdcopy.StdCopy(os.Stdout, os.Stderr, reader)
	if err != nil {
		panic(err)
	}
}
```
</details>

```
$ ./test -n 2
stdout 10                  #  <---- only the last line that cames from stdout was returned

$ ./test -n 3
stdout 8
stdout 10
```

knight42 commented:
/remove-lifecycle rotten

knight42 committed:
run "make update"

Signed-off-by: Jian Zeng <anonymousknight96@gmail.com>

knight42 committed:
style: make linter happy

Signed-off-by: Jian Zeng <anonymousknight96@gmail.com>

APErebus commented:
What's the status of this PR? This would be very helpful and it looks like it's stagnant? Is this awaiting the author just for a rebase?

knight42 commented:
/reopen

knight42 commented:
Since there is too much merge conflict in this PR, I will open a new one.
